shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -184,6 +184,152 @@ abstract class JSBackend(allowUnresolvedSymbols: Bool) {
+      createASTCall("Var", StrLit("error") :: Nil)
+    case Var(name) if isQuoted || freeVars.vs(name) => createASTCall("Var", Var(scope.resolveValue(name).fold[Str](
+      throw CodeGenError(s"unbound free variable $name is not supported yet.")
+    )(_.runtimeName)) :: Nil)
+    case lit: IntLit if isQuoted => createASTCall("IntLit", lit :: Nil)
+    case lit: DecLit if isQuoted => createASTCall("DecLit", lit :: Nil)
+    case lit: StrLit if isQuoted => createASTCall("StrLit", lit :: Nil)
+    case lit: UnitLit if isQuoted => createASTCall("UnitLit", lit :: Nil)
+    case Lam(params, body) =>
+      if (isQuoted) {
+        val lamScope = scope.derive("Lam")
+        params match {
+          case Tup(params) =>
+            val newfreeVars = params.map {
+              case N -> Fld(_, Var(nme)) =>
+                lamScope.declareParameter(nme)
+                nme -> lamScope.declareValue(nme, S(false), false, N).runtimeName
+              case S(Var(nme)) -> _ =>
+                lamScope.declareParameter(nme)
+                nme -> lamScope.declareValue(nme, S(false), false, N).runtimeName
+              case p => throw CodeGenError(s"parameter $p is not supported in quasiquote")
+            }
+            newfreeVars.foldRight(desugarQuote(body)(lamScope, isQuoted, new FreeVars(freeVars.vs ++ newfreeVars.map(_._1))))((p, res) =>
+              Let(false, Var(p._2), createASTCall("freshName", StrLit(p._1) :: Nil), createASTCall("Lam", createASTCall("Var", Var(p._2) :: Nil) :: res :: Nil)))
+          case _  => throw CodeGenError(s"term $params is not a valid parameter list")
+        }
+      }
+      else Lam(params, desugarQuote(body))
+    case Unquoted(body) if isQuoted =>
+      val unquoteScope = scope.derive("unquote")
+      desugarQuote(body)(unquoteScope, false, freeVars)
+    case Quoted(body) =>
+      val quoteScope = scope.derive("quote")
+      val res = desugarQuote(body)(quoteScope, true, freeVars)
+      if (isQuoted) createASTCall("Quoted", res :: Nil)
+      else res
+    case App(App(Var(op), Tup((N -> Fld(f1, lhs)) :: Nil)), Tup((N -> Fld(f2, rhs)) :: Nil))
+        if JSBinary.operators contains toJSOperator(op) =>
+      if (isQuoted)
+        createASTCall("App", createASTCall("Var", StrLit(toJSOperator(op)) :: Nil) :: desugarQuote(lhs) :: desugarQuote(rhs) :: Nil)
+      else
+        App(App(Var(toJSOperator(op)), Tup((N -> Fld(f1, desugarQuote(lhs))) :: Nil)), Tup((N -> Fld(f2, desugarQuote(rhs))) :: Nil))
+    case App(Var(op), Tup(N -> Fld(f1, lhs) :: N -> Fld(f2, rhs) :: Nil))
+        if JSBinary.operators.contains(toJSOperator(op)) && (!translateVarImpl(op, isCallee = true).isRight || op =/= toJSOperator(op)) =>
+      if (isQuoted)
+        createASTCall("App", createASTCall("Var", StrLit(toJSOperator(op)) :: Nil) :: desugarQuote(lhs) :: desugarQuote(rhs) :: Nil)
+      else
+        App(Var(toJSOperator(op)), Tup(N -> Fld(f1, desugarQuote(lhs)) :: N -> Fld(f2, desugarQuote(rhs)) :: Nil))
```

Comment:
Why do you try to translate JS binary operators here? This logic is already in `translateTerm`.